---
title: "MOFA analysis of the Chromium Single Cell Multiome ATAC + Gene Expression assay"
author:
- name: "Ricard Argelaguet"
  affiliation: "European Bioinformatics Institute, Cambridge, UK"
  email: "ricard@ebi.ac.uk"
- name: "Danila Bredikhin"
  affiliation: "European Molecular Biology Laboratory, Heidelberg, Germany"
- name: "Oliver Stegle"
  affiliation: "German Cancer Research Center(DKFZ), Heidelberg, Germany"
- name: "John Marioni"
  affiliation: "Cancer Research UK, Cambridge, United Kingdom"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Description

This vignette demonstrates how MOFA can be used to integrate scRNA-seq and scATAC-seq data from the [Chromium Single Cell Multiome ATAC + Gene Expression assay](https://support.10xgenomics.com/single-cell-multiome-atac-gex/) recently commercialised by 10x Genomics. This vignette results from a collaboration between the 10x Genomics R&D team and the MOFA team. The data set consists of the conventional Peripheral Blood Mononuclear Cells (PBMC) from a single healthy donor, which is available [here](https://support.10xgenomics.com/single-cell-multiome-atac-gex/datasets).

[MOFA is a factor analysis model that provides a general framework for the integration of multi-omic data sets in an unsupervised fashion](https://www.embopress.org/doi/10.15252/msb.20178124).
Intuitively, MOFA can be viewed as a versatile and statistically rigorous generalisation of principal component analysis (PCA) to multi-omics data. Briefly, MOFA performs unsupervised dimensionality reduction simultaneously across multiple data modalities, thereby capturing the global sources of cell-to-cell variability via a small number of inferred factors. Importantly, MOFA distinguishes variation that is shared between assays from variation that is unique to a specific assay. Thus, in this data set MOFA can be useful to disentangle the RNA and the ATAC activity of the different cellular populations that exist in PBMCs.

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE, error=FALSE, fig.align="center")
```

# Load libraries

Make sure that `MOFA2` is imported last, to avoid collisions with functions from other packages
```{r, message=FALSE}
library(data.table)
library(ggplot2)
library(Seurat)
library(Signac)

# for GSEA analysis
library(msigdbr)

# For motif enrichment analysis
library(JASPAR2020)
library(TFBSTools)
library(BSgenome.Hsapiens.UCSC.hg38)

# MOFA
library(MOFA2)
```

```{r echo=FALSE}
matrix.please<-function(x) {
  m<-as.matrix(x[,-1])
  rownames(m)<-x[[1]]
  m
}
```

# Load data

## Load multi-modal Seurat object

We have created a Seurat object with RNA and ATAC data modalities stored as different assays. [Here is the script](ftp://ftp.ebi.ac.uk/pub/databases/mofa/10x_rna_atac_vignette/create_seurat.R) to create the Seurat object from the CellRanger output
```{r}
# run this to download the data set
# seurat <- readRDS(url("ftp://ftp.ebi.ac.uk/pub/databases/mofa/10x_rna_atac_vignette/seurat.rds"))

seurat <- readRDS("/Users/ricard/data/10x_rna_atac/PBMC/seurat.rds")

seurat
```

The metadata slot contains the cell type annotations that have been done *a priori* by the 10x Genomics R&D team. This will be useful to characterise the MOFA factors. One could employ the MOFA factors to perform clustering and cell type annotation, but we will demonstrate this in a future vignette.
```{r}
head(seurat@meta.data[,c("celltype","broad_celltype","pass_rnaQC","pass_accQC")])
```

```{r}
table(seurat@meta.data$celltype)
```

```{r}
table(seurat@meta.data$broad_celltype)
```

Keep cells that pass QC for both omics
```{r}
seurat <- seurat %>%
  .[,seurat@meta.data$pass_accQC==TRUE & seurat@meta.data$pass_rnaQC==TRUE]
seurat
```

The RNA expression consists of 29,732 genes and 10,032 cells
```{r}
seurat@assays[["RNA"]]
```

The ATAC expression consists of 108,377 peaks and 10,032 cells
```{r}
seurat@assays[["ATAC"]]
```

## Load additional information

Collect a list of position-specific weight matrices (PWM) from the JASPAR database, we'll use this in the downstream analysis.
```{r}
pfm <- getMatrixSet(JASPAR2020,
  opts = list(species = "Homo sapiens")
)
```

```{r echo=FALSE}
mapping <- fread("ftp://ftp.ebi.ac.uk/pub/databases/mofa/10x_rna_atac_vignette/JASPAR2020_mapping.txt")
# mapping <- fread("/Users/ricard/data/JASPAR/JASPAR2020_mapping.txt")
stopifnot(names(pfm) %in% mapping$id)
foo <- mapping$name; names(foo) <- mapping$id
names(pfm) <- foo[names(pfm)] %>% paste0(.,"-motif")
```

# Load metadata

## Cell metadata


```{r, echo=FALSE}
seurat$celltype[seurat$celltype=="MAIT T cells"] <- "effector CD8 T cells"

celltype_levels <- c(
	"naive CD4 T cells", 
	"memory CD4 T cells", 
	"naive CD8 T cells", 
	# "MAIT T cells", 
	"non-classical monocytes", 
	"classical monocytes", 
	"effector CD8 T cells", 
	"myeloid DC", 
	"plasmacytoid DC", 
	"intermediate monocytes", 
	"naive B cells", 
	"memory B cells", 
	"CD56 (bright) NK cells", 
	"CD56 (dim) NK cells"
)
seurat$celltype <- factor(seurat$celltype,levels=celltype_levels)

gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}
colors <- gg_color_hue(length(celltype_levels))
names(colors) <- celltype_levels
```

## Feature metadata

Feature metadata contains genomic information for both RNA expression and ATAC peaks
```{r}
# feature_metadata <- fread("/Users/ricard/data/10x_rna_atac/PBMC/original/filtered_feature_bc_matrix/features.tsv.gz") %>%
feature_metadata <- fread("ftp://ftp.ebi.ac.uk/pub/databases/mofa/10x_rna_atac_vignette/filtered_feature_bc_matrix/features.tsv.gz") %>%
  setnames(c("ens_id","gene","view","chr","start","end"))
```

Fetch RNA metadata
```{r}
feature_metadata.rna <- feature_metadata[view=="Gene Expression"]
head(feature_metadata.rna,n=3)
```

Fetch ATAC metadata and classify peaks into promoter-overlapping and distal
```{r}
feature_metadata.atac <- feature_metadata[view=="Peaks"] %>% 
  .[,ens_id:=NULL] %>% setnames("gene","peak")
head(feature_metadata.atac,n=3)
```

```{r, echo=FALSE}
# foo <- fread("ftp://ftp.ebi.ac.uk/pub/databases/mofa/10x_rna_atac_vignette/atac_peak_annotation.tsv") %>%
foo <- fread("/Users/ricard/data/10x_rna_atac/PBMC/original/old/atac_peak_annotation.tsv") %>%
  .[,c("peak","peak_type")] %>%
  .[peak_type%in%c("distal", "promoter")]
feature_metadata.atac <- feature_metadata.atac %>% merge(foo,by="peak",all.x=TRUE)
```

```{r}
table(feature_metadata.atac$peak_type)
```

# Parse Seurat object

<!-- Remove non-conventional chromosomes -->
```{r, echo=FALSE}
feature_metadata.atac <- feature_metadata.atac %>%
  .[chr%in%paste0("chr",c(1:22,"X"))]
```

Split ATAC matrix depending on the peak type and create a `ChromatinAssay` for each modality using the [Signac package](https://satijalab.org/signac). This object requires a GRanges object with the peak metadata. We also provide an optional motif matrix that we'll use for downstream analysis.
```{r, message=FALSE, warning=FALSE}
for (i in c("distal","promoter")) {
  
  # Create GRanges
  peaks.granges <- feature_metadata.atac %>%
    .[peak_type==i] %>%
    .[,c("chr","start","end","peak")] %>%
    makeGRangesFromDataFrame(keep.extra.columns = TRUE, ignore.strand = TRUE)

  # Scan motifs throughout the DNA sequence of each peak and create a binary matrix of motif-peak presence.
  motif.matrix <- CreateMotifMatrix(
    features = peaks.granges,
    pwm = pfm,
    genome = 'hg38',
    use.counts = FALSE
  ) %>% as.matrix
  
  # AddChromatinAssay to the Seurat object
  seurat@assays[[paste0("ATAC_",i)]] <- CreateChromatinAssay(
    seurat@assays$ATAC@counts[peaks.granges$peak,], 
    ranges = peaks.granges,
    motifs = CreateMotifObject(motif.matrix, pfm)
  )
  
}
seurat
```

```{r echo=FALSE}
# for (i in c("ATAC_distal","ATAC_promoter")) {
#   dim(seurat[[i]]) %>% print
#   length(seurat[[i]]@ranges) %>% print
#   dim(seurat[[i]]@motifs) %>% print
# }
```

# Normalization

## RNA

Standard Log normalisation, nothing fancy. One could possibly regress out the `nFeature_RNA` covariate or use the alternative `SCTransform` normalisation
```{r}
seurat <- NormalizeData(seurat, normalization.method = "LogNormalize", assay = "RNA")
seurat <- ScaleData(seurat, do.center = TRUE, do.scale = FALSE)
```

## ATAC

TFIDF normalisation, as implemented in `Signac`. This is important to transform the binary measurements to continuous readouts, which is a mandatory step to use the Gaussian likelihood in MOFA. Keep in mind that other strategies are possible (and perhaps even more recommended). One could also try to scale the data and regress out `nFeature_ATAC`
```{r}
for (i in c("ATAC_distal","ATAC_promoter")) {
  seurat <- RunTFIDF(seurat, assay = i)
}
```

# Feature selection

## RNA

Select most variable genes using the VST procedure
```{r message=FALSE }
seurat <- FindVariableFeatures(seurat, 
  selection.method = "vst", 
  nfeatures = 5000,
  assay = "RNA",
  verbose = FALSE
)
```

## ATAC

Select the most variable peaks using the `FindTopFeatures` function
```{r}
for (i in c("ATAC_distal","ATAC_promoter")) {
  seurat <- FindTopFeatures(seurat, assay=i, min.cutoff = 2000)
  print(length(seurat[[i]]@var.features))
}
```

# Train the MOFA model

## Create MOFA object

MOFA can take as input a multi-assay Seurat object. 
<!-- Notice that [MOFA now enables simultaneous multi-group and multi-modal analysis ](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02015-1). In this data set we do not have a clear grouping structure (i.e batches, donors, time points, etc.), so we will use the default option of not doing multi-group inference. -->
```{r}
mofa <- create_mofa(seurat, assays = c("RNA","ATAC_distal","ATAC_promoter"))
mofa
```

## Define MOFA options

The most important hyperparameter is the number of factors. It needs to be specified by the user and it generally depends on the complexity of the data set. It is usually recommended to provide a large enough number and trim factors _a posteriori_ based on a minimum variance explained criteria. Here let's define K=15 factors
```{r, message=FALSE}
model_opts <- get_default_model_options(mofa)
model_opts$num_factors <- 15
```

There also other options that can be tuned, see `?get_default_model_options`, `?get_default_training_options` and `?get_default_data_options` for details.
```{r, message=FALSE, warning=FALSE}
mofa <- prepare_mofa(mofa,
  model_options = model_opts
)
```

## Run MOFA

This step can take around 2h in a MacBook Pro, but it can be speed up using multiple cores (~2-3x) or using GPUs (~10x). The training is done in Python, and the connection with R is done via [reticulate](https://rstudio.github.io/reticulate/). If this step fails please read our [FAQ](https://github.com/bioFAM/MOFA2/#4-faq-on-the-software), as this is the most common source of errors when running MOFA
```{r message=FALSE}
# mofa <- run_mofa(mofa)
```

Load pretrained MOFA object
```{r}
# run this to download the data set
# mofa <- readRDS(url("ftp://ftp.ebi.ac.uk/pub/databases/mofa/10x_rna_atac_vignette/mofa.rds"))

mofa <- readRDS("/Users/ricard/data/10x_rna_atac/PBMC/mofa/mofa.rds")

mofa
```

# MOFA downstream analysis

## Add cell metadata to the model

The sample metadata must be provided as a data.frame and it must contain a column named _sample_ with the sample IDs.

```{r}
samples_metadata(mofa) <- seurat@meta.data %>%
  tibble::rownames_to_column("sample") %>%
  as.data.table
```

## Asses the correlation between factors

A good sanity check is to verify that the Factors are largely uncorrelated. In MOFA there are no orthogonality constraints such as in Principal Component Analysis, but if there is a lot of correlation between Factors this suggests a poor model fit
```{r}
plot_factor_cor(mofa)
```

## Variance decomposition

The variance decomposition analysis is my favourite MOFA functionality. It calculates the percentage of variance explained by each factor and data modality. What insight can we learn just from visualising this? Factor 1 has a very strong signal from all data modalities, so it must be the most important source of variability in the data. Factor 2 and Factor 3 are weaker but also shared between all data modalities, so they are also likely to be impotant sources of variability. Interestingly, most Factors have a stronger signal in distal peaks than in promoter peaks.

```{r}
plot_variance_explained(mofa, max_r2 = 4)
```

A reasonable question is whether the model is providing a good fit to the data. To assess this we can plot the total variance explained (using all factors). The resulting values will depend on the nature of the data set, the number of samples, the number of factors, etc. For single-cell data the values tend to be quite low. Here we have a 30% for RNA, a very reasonable value, versus ~8-10% for ATAC data, which is also a decent value taking into account how noisy scATAC data is. In fact, compared to my previous analysis on [the SNARE-seq data set](https://raw.githack.com/bioFAM/MOFA2/master/MOFA2/vignettes/SNARE_seq.html#variance-decomposition-1) these are very good values. In the SNARE-seq analysis the 
the ATAC modality was extremely noisy and we had to rely on denoising strategies to get meaningful signal. As you'll see in this vignette, the 10x Genomics technology provides much better quality measurements and no denoising strategy is required.

```{r}
plot_variance_explained(mofa, plot_total = TRUE)[[2]]
```

## Characterisation of Factors

There are a few systematic strategies to characterise the molecular etiology underlying each MOFA Factor:

- Association analysis between cell covariates and Factor values. 
- Visualisation of factor values.
- Visualisation of feature weights. 
- Gene set enrichment analysis. 

### Correlate factors with covariates

As a first approach, let's correlate the standard quality control metrics with the Factor values. We see that Factor 4 is strongly associated with the number of expressed genes per cell and the Factors 1 and 7 are associated with the number of accessible peaks per cell. Some of these are likely technical factors that should be removed with a better normalisation strategy.
```{r}
correlate_factors_with_covariates(mofa, 
  covariates = c("nFeature_RNA","nFeature_ATAC")
  )
```

### Characterisation of Factor 1

The variance expalined plot tells us that Factor 1 is the strongest source of variation and has a large contribution from all data modalities.

#### Visualisation of Factor values

**How do we interpret the factor values?**  
Mathematically, each Factor is defined by a linear combination of the input features. As the data is centered prior to running MOFA, each Factor ordinates cells along a one-dimensional axis that is centered at zero. Samples with different signs manifest opposite "effects" along the inferred axis of variation, with higher absolute value indicating a stronger effect. Note that the interpretation of MOFA factors is analogous to the interpretation of the principal components in PCA.

Let's plot the Factor 1 values grouped by cell type and coloured by lineage of origin (Lymphoid or Myeloid). Clearly Factor 1 captures the molecular variation associated with lineage of origin.
```{r}
plot_factor(mofa, factors=1, group_by = "celltype", color_by="broad_celltype") +
  theme(
    axis.text.x = element_text(color="black", angle=40, vjust=1, hjust=1)
  )
```

#### Visualisation of feature weights

**how do we interpret the weights?**  
The weights provide a score for each feature on each factor. Features with no association with the corresponding factor are expected to have values close to zero, whereas features with strong association with the factor are expected to have large absolute values. The sign of the weights indicates the direction of the effect: a positive weights indicates that the feature has higher levels in the cells with positive factor values, and vice-versa.

Let's plot the associated RNA weights. These are all good markers of myeloid (LYN, SLC8A1, LYZ, etc.) vs lymphoid fate (BCL11B)
```{r}
plot_weights(mofa, 
  view = "RNA", 
  factors = 1, 
  nfeatures = 20, 
  text_size = 4
)
```

#### Visualisation covariation patterns in the input data

The weights are useful to get an idea of which genes are driving each factor. However, to get an idea of how good the association between features and factors is we can generate a scatterplot of the Factor values against mRNA expression for the genes with the largest weights:
```{r}
plot_data_scatter(mofa, 
  view = "RNA", 
  factor = 1, 
  features = 6,
  color_by = "broad_celltype",
  add_lm = T,
  dot_size = 1
)
```

Now let's visualise the ATAC signatures by inspecting the weights, although they are not very interpretable. We'll have to rely on motif-based approaches to get TF/motif signatures.
```{r}
plot_top_weights(mofa, 
  view = "ATAC_distal", 
  factors = 1, 
  sign = "positive",
  nfeatures = 15,
)
```

Again, we can visualise the ATAC patterns recovered by MOFA on the high-dimensional space. Instead of scatterplots we can use heatmaps:
```{r }
plot_data_heatmap(mofa, 
  view = "ATAC_promoter", 
  factor = 1, 
  features = 50,
  show_rownames = F, show_colnames = F, 
  cluster_rows = T, cluster_cols = F,
  annotation_samples = "broad_celltype"
)
```

Again, we can visualise the ATAC patterns recovered by MOFA on the high-dimensional space. Instead of scatterplots we can also use heatmaps:

MOFA has an interesting option to denoise/smooth the data by reconstructing the data modalities using the latent factors. We can check how this works in the `plot_data_heatmap` function using `denoise = TRUE`. Instead of plotting the (noisy) input data, this plots the data reconstructed by the model:
```{r }
plot_data_heatmap(mofa, 
  view = "ATAC_promoter", 
  factor = 1, 
  features = 50,
  show_rownames = F, show_colnames = F, 
  cluster_rows = T, cluster_cols = F,
  annotation_samples = "broad_celltype",
  denoise = TRUE
)
```

### Characterisation of Factor 2

Let's proceed now in a similar way to characterise Factor 2. It is clearly capturing B cell commitment

```{r}
plot_factor(mofa, factors=2, group_by = "celltype") +
  theme(
    axis.text.x = element_text(color="black", angle=40, vjust=1, hjust=1)
  )
```

Let's plot the associated RNA weights. We see good B cell markers, including CD74 and BANK1
```{r}
plot_weights(mofa, 
  view = "RNA", 
  factors = 2, 
  nfeatures = 10, 
  text_size = 4
)
```

Let's color the factor values by BANK1 expression
```{r}
plot_factor(mofa, factors=2, group_by = "celltype", color_by="BANK1") +
  theme(
    axis.text.x = element_text(color="black", angle=40, vjust=1, hjust=1)
  )
```

Again, we can visualise the ATAC signatures:
```{r }
plot_data_heatmap(mofa, 
  view = "ATAC_promoter", 
  factor = 2, 
  features = 50,
  show_rownames = F, show_colnames = F, 
  cluster_rows = T, cluster_cols = F,
  annotation_samples = "celltype",
  denoise = TRUE
)
```

### Characterisation of Factor 4

In the correlation analysis we found that Factor 4 is strongly correlated to the total number of ATAC peaks, so it is likely a technical source of variation that has not been removed in the normalisation steps. The signal is clear, and it affects all cell types in a similar fashion:

```{r}
plot_factor(mofa, factors=4, group_by = "celltype", color_by="nFeature_ATAC") +
  theme(
    axis.text.x = element_text(color="black", angle=40, vjust=1, hjust=1)
  )
```

If we plot the associated ATAC weights notice that the distribution is massively shifted towards negative values. A typical signature of a  technical factor is that most of the weights tend to non-zero:
```{r}
plot_weights(mofa, 
  view = "ATAC_promoter", 
  factors = 4, 
  nfeatures = 0, 
  text_size = 4
)
```

## Non-linear dimensionality reduction

### Using MOFA factors

The latent space inferred by MOFA can replace the PCA embedding as input to algorithms that learn non-linear manifolds such as t-SNE or UMAP. This can be very useful to identify cellular populations and reconstruct complex pseudotime trajectories. The advantage of  MOFA is that (1) we use information from all available omics, and (2) we can characterise the Factors and remove the technical ones.

```{r message=FALSE}
factors <- 1:get_dimensions(mofa)[["K"]]
factors <- factors[!factors%in%c(4,7)]

mofa <- run_umap(mofa, 
  factors = factors, 
  n_neighbors = 15,  
  min_dist = 0.30
)

plot_dimred(mofa, 
  method = "UMAP", 
  color_by = "celltype", 
  label = TRUE, 
  stroke=0.05, 
  dot_size = 1, 
  legend = FALSE
) + scale_fill_manual(values=colors)
```

We can try to add some interpretatibility on the UMAP by visualising the contribution of each Factor on the different groups of cells.
```{r}
for (i in paste0("Factor",1:3)) {
  p <- plot_dimred(mofa, 
    method = "UMAP", 
    color_by = i, 
    stroke = 0.05, 
    dot_size = 1
  )
  print(p)
}
```

In this case however we notice that the resulting UMAP is not very different from the one you could obtain using the RNA expression or the ATAC data alone. This is because most, if not all, biological factors have contributions from both the RNA and ATAC modalities.

```{r echo=FALSE}
seurat <- SetIdent(seurat, value="celltype")
K <- get_dimensions(mofa)[["K"]]
```

### Using RNA data alone

```{r message=FALSE}
DefaultAssay(seurat) <- "RNA"
seurat <- RunPCA(seurat, npcs = K, verbose = FALSE)
seurat <- RunUMAP(seurat, reduction = 'pca', dims = 1:K, verbose = FALSE)
DimPlot(seurat, label = TRUE, reduction="umap") + 
  NoLegend() + NoAxes() + scale_fill_manual(values=colors)
```

### Using ATAC data alone

```{r message=FALSE}
DefaultAssay(seurat) <- "ATAC_distal"
seurat <- RunSVD(seurat, n = K, verbose = FALSE)
seurat <- RunUMAP(seurat, reduction = 'lsi', dims = 1:K, verbose = FALSE)
DimPlot(seurat, label = TRUE, reduction="umap") + 
  NoLegend() + NoAxes() + scale_fill_manual(values=colors)
```

## Gene set enrichment analysis

In addition to exploring the individual weights for each factor, we can use enrichment analysis to look for significant associations of factors to genesets. For more details on how GSEA works in MOFA we encourage the users to read the [GSEA vignette](https://raw.githack.com/bioFAM/MOFA2/master/MOFA2/vignettes/GSEA.html).

### Load gene set annotations

First we define the gene set matrix. We'll use the C5 category and the Biological Process subcategory from the [MSigDB data base](https://www.gsea-msigdb.org/gsea/msigdb):
```{r}
msgidb.matrix <- msigdbr(
  species = "Homo sapiens",
  category = "C5", 
  subcategory = "BP"
  ) %>% as.data.table %>% .[,id:=1] %>%
  dcast(gs_name~gene_symbol, value.var="id", fill=0) %>% 
  matrix.please
```


### Run GSEA

Second, we run the enrichment analysis with default options. An important consideration when running GSEA is that MOFA Factors contain positive (+) and negative (-) weights. There will be cases where the genes with (-) weights all belong to a specific pathway but genes with (+) weights belong to other pathways. If this is true, doing GSEA with all of them together could dilute the signal. Hence, we recommend the user to do GSEA separately for (+) and (-) weights.

```{r message=FALSE}
# GSEA on positive weights
gsea.positive <- run_enrichment(mofa, 
  feature.sets = msgidb.matrix, 
  view = "RNA",
  sign = "positive"
)

# GSEA on negative weights
gsea.negative <- run_enrichment(mofa, 
  feature.sets = msgidb.matrix, 
  view = "RNA",
  sign = "negative"
)
```

The enrichment analysis returns a list of 5 elements:  

- **feature.sets**:	the gene set matrix filtered by the genes that overlap with the MOFA model.  
- **pval**:	the nominal p-values.  
- **pval.adj**:	the FDR-adjusted p-values.  
- **feature.statistics**: the feature statistics (i.e. the mofa weights, by default).  
- **set.statistics**: the gene set statistics.  
- **sigPathways**: list with significant pathways per factor at a specified FDR threshold  

```{r}
names(gsea.positive)
```

### Visualise GSEA rsults

Factor 1 captured lineage of origin, where lymphoid cells wer linked to (-) factor values and myeloid cells to (+) factor values. Consistently, when plotting the pathways enriched in Factor 1 for (+) weight values we can see Myeloid-associated pathways:
```{r}
plot_enrichment(gsea.positive, factor = 1, max.pathways = 15)
```

and when plotting the pathways enriched in Factor 1 for (-) weight values we can see Lymphoid-associated pathways:
```{r}
plot_enrichment(gsea.negative, factor = 1, max.pathways = 15)
```

It is always advised to not rely only on the p-values and to visualise which genes are driving the enrichment within each pathways. This can be done with the `plot_enrichment_detailed` function.
```{r}
plot_enrichment_detailed(gsea.positive,
  factor = 1,
  max.genes = 10,
  max.pathways = 5
)
```

## Motif enrichment 

Although peak annotation provides functional interpretation, it does not directly explain the underlying mechanism. Open chromatin can affect transcription through TFs, which facilitate transcription by recognizing and binding to specific DNA motifs. Let's attempt to do motif enrichment analysis using the weights of the ATAC modality.

Here we can use the GSEA functionality implemented in MOFA. We just replace the gene set annotations by peak-motif annotations. 

### Run

The variance explained plot suggests that there is more signal in distal peaks than in promoter peaks in the first two factors. Let's do motif enrichment on distal peaks. Again, we compare positive weights (for Factor 1, peaks more open and motifs more enriched in Lymphoid cells) versus negative weights (for Factor 1, peaks more open and motifs more enriched in Myeloid cells)

```{r echo=FALSE}
# small tweak to match the peak names
features_names(mofa)[["ATAC_distal"]] <- features_names(mofa)[["ATAC_distal"]] %>% stringr::str_replace_all(":","-")
```

```{r message=FALSE}
# define motif matrix
motif.matrix <- t(as.matrix(seurat[["ATAC_distal"]]@motifs@data))

# Run GSEA enrichment analysis using the motif-peak matrix, (+) weights
motif.enrichment.positive <- run_enrichment(mofa,
  view = "ATAC_distal", 
  factors = 1:2,
  feature.sets = motif.matrix,
  sign = "positive"
)

# Run GSEA enrichment analysis using the motif-peak matrix, (-) weights
motif.enrichment.negative <- run_enrichment(mofa,
  view = "ATAC_distal", 
  factors = 1:2,
  feature.sets = motif.matrix,
  sign = "negative"
)
```

### Visualise

Plot motif enrichment results for Factor 1
```{r}
plot_enrichment(motif.enrichment.positive, factor = 1, max.pathways = 15)
plot_enrichment(motif.enrichment.negative, factor = 1, max.pathways = 15)
```
We can visualise the motifs using the `MotifPlot` function from the `Signac` package.

Myeloid motifs (positive Factor values): `CEBP`s are [one of the most studied and important regulator of myeloid differentiation](https://ashpublications.org/blood/article/129/15/2083/36382/Expression-and-regulation-of-C-EBP-in-normal).

```{r}
sig.motifs.positive <- motif.enrichment.positive$pval.adj[,"Factor1"] %>%
  sort %>% head(n=6) %>% names
MotifPlot(seurat[["ATAC_distal"]], motifs = sig.motifs.positive)
```

Lymphoid motifs (negative Factor values):  

- `LEF1` [is essential for T-cell functionality](https://en.wikipedia.org/wiki/Lymphoid_enhancer-binding_factor_1). 
- `TCFL2` also known as TCF4, is a [T-cell specific TF](https://en.wikipedia.org/wiki/TCF7L2). 
- `RUNX3` is [highly expressed in cells of lymphoid origin](https://en.wikipedia.org/wiki/RUNX3#/media/File:PBB_GE_RUNX3_204197_s_at_fs.png). 
- `TCF7` is [highly expressed in cells of lymphoid origin](https://en.wikipedia.org/wiki/TCF7#/media/File:PBB_GE_TCF7_205254_x_at_fs.png). 

Seems like we are getting sensible results!
 
```{r}
sig.motifs.negative <- motif.enrichment.negative$pval.adj[,"Factor1"] %>%
  sort %>% head(n=6) %>% names
MotifPlot(seurat[["ATAC_distal"]], motifs = sig.motifs.negative)
```

### Validation using chromVAR

To validate the motif enrichment results, we can calculate motif activity scores per cell using [chromVAR](https://bioconductor.org/packages/release/bioc/html/chromVAR.html). Briefly, this method computes an accessibility z-score for each motif in each cell that is adjusted by technical confounders. Positive values denote more accessible than background, and negative values denote less accessible than background

```{r}
seurat <- RunChromVAR(
  object = seurat, 
  assay = "ATAC_distal",
  genome = BSgenome.Hsapiens.UCSC.hg38, 
  new.assay.name = "chromvar"
)

DefaultAssay(seurat) <- "chromvar"
```

Indeed, the top hits for Factor 1 have very clear chromVAR patterns that distinguish lymphoid versus myeloid cells
```{r message=FALSE}
motifs.to.plot <- c(sig.motifs.positive[1:2], sig.motifs.negative[1:2])

FeaturePlot(seurat,
  features = motifs.to.plot,
  reduction = "umap",
  combine = TRUE
) & NoLegend() & NoAxes()
```

# Conclusions

This vignette demonstrates how MOFA can be used to integrate the RNA and ATAC modalities from the Chromium Single Cell Multiome ATAC + Gene Expression assay recently introduced by 10x Genomics. I want to thank Vijay, Shamoni, Paul and Kamila from the 10x R&D team for this collaboration and for feedback on the analysis. 

If you have questions, suggestions about MOFA or the vignette you can contact me by email (ricard@ebi.ac.uk). Also, we have a [Slack group](https://join.slack.com/t/mofahelp/shared_invite/enQtMjcxNzM3OTE3NjcxLWNhZmM1MDRlMTZjZWRmYWJjMGFmMDkzNDBmMDhjYmJmMzdlYzU4Y2EzYTI1OGExNzM2MmUwMzJkZmVjNDkxNGI) where we provide quick and personalised advise if you are interested in MOFA.

<details>
  <summary>**Session Info**</summary>
  
```{r}
sessionInfo()
```

</details>
