---
title: " Vignette illustrating the use of MOFA on the CLL data"
author: "Britta Velten and Ricard Argelaguet"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{Example_CLL}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

This vignette show how to use MOFA including initialization, training and down-stream analysis.
For illustration we use the CLL data which is used in the MOFA publication.

```{r, warning=FALSE, message=FALSE}
library(MOFAtools)
```

# Step 1: Load data and create MOFA object
There are two options to input data to MOFA:
- Option 1: base R approach using a list of matrices
- Option 2: Bioconductor approach using the MultiAssayExperiment framework

## Option 1: base R approach
If using the base R approach, you simply need to provide a list of matrices where features are rows and samples are columns. Importantly, the samples need to be aligned. Missing values/assays should be filled with NAs.
```{r}
data("CLL_data")
MOFAobject <- createMOFAobject(CLL_data)
MOFAobject
```

## Option 2: bioconductor approach
If using the Bioconductor approach, you need to provide or create a [MultiAssayExperiment](https://bioconductor.org/packages/release/bioc/html/MultiAssayExperiment.html) object and then use it to build the MOFA object. For example, starting from a list of matrices where features are rows and samples are columns, this can be easily constructed as follows:
```{r}
library(MultiAssayExperiment)

# Load data
data("CLL_data") # import list containing matrices for mRNA, Methylation, Drug Response and Mutation data. 
lapply(CLL_data, dim) # check dimensionalities, samples are columns, features are rows

# Load sample metadata (colData): Sex and Diagnosis
data("CLL_covariates")
head(CLL_covariates)

# Create MultiAssayExperiment object 
mae_CLL <- MultiAssayExperiment(experiments = CLL_data, colData = CLL_covariates)

# Build the MOFA object
MOFAobject <- createMOFAobject(mae_CLL)
MOFAobject

```


## Overview of training data
The function 'plotTilesData' can be used to obtain an overview of the data stored in the object for training. FOr each sample it shows in which views data are available.
```{r}
plotTilesData(MOFAobject)
```

# Step 2: Fit the MOFA model
The next step is to fit the model. 
This part of the pipeline is implemented in Python, so first of all make sure you have the corresponding package installed (see installation instructions in the README file). In principle, we recommend to train the model directly from the command line (see running instructions, it is very simple), but we have created the following R wrapper if you do not feel comfortable with bash.

## Define options

### Define I/O options
```{r}
DirOptions <- list(
  "dataDir" = tempdir(), # Folder to store the input matrices as .txt files, it can be a simple temporary folder
  "outFile" = "/Users/ricard/test/test.hdf5" # Output file of the model (use hdf5 extension)
)
```

### Define model options
The most important options the user needs to define are:
* numFactors: initial number of factors
* likelihoods: likelihood for each view, gaussian for continuous data, bernoulli for binary data and poisson for count data. If not provided, the model tries to guess it from the data.
```{r}
ModelOptions <- getDefaultModelOpts(MOFAobject)
ModelOptions
```

### Define training options
The most important options the user needs to define are:
* maxiter: Maximum number of iterations, ideally set it large enough and use the convergence criteria
* tolerance: Convergence threshold based on change in the evidence lower bound. We recommend you stick to the default value.
* DropFactorThreshold: Threshold on fraction of variance explained to define inactive factors. That is, factors explaining less than 'DropFactorThreshold' percentage of variation in all views will be dropped during training.
```{r}
TrainOptions <- getDefaultTrainOpts()
TrainOptions$maxiter <- 5
TrainOptions
```

## Prepare MOFA
prepareMOFA internally performs a set of sanity checks, fills the TrainOpts and ModelOpts slots of the MOFAmodel object and it also creates a set of temporary files with the input matrices that will be loaded by the Python core implementation
```{r}
MOFAobject <- prepareMOFA(MOFAobject, 
                      DirOptions = DirOptions,
                      ModelOptions = ModelOptions,
                      TrainOptions = TrainOptions
)
```

## Run MOFA
This step can take some time (around 20 min with default parameters), for illustration we provide an existing trained model
```{r, eval=FALSE}
# MOFAobject <- runMOFA(MOFAobject, DirOptions)
```

```{r}
# Loading an existing trained model
filepath <- system.file("extdata", "model15.hdf5", package = "MOFAtools")
MOFAobject <- loadModel(filepath, MOFAobject)
MOFAobject
```


# Step 3: Analyse a trained MOFA model
After training, we can explore the results from MOFA. Here we provide a semi-automated pipeline to disentangle and characterize the sources of variation (the factors) identified by MOFA.

Part 1: Disentangling the heterogeneity:
- Calculation of variance explained by each factor in each view
Part 2: Characterisation of individual factors:
- Inspection of top weighted features in the active views
- Feature set enrichemnt analysis in the active views (where set annotations are present, e.g. gene sets for mRNA views)
- Ordination of samples by factors to reveal clusters and graadients in the sample space

For details, please read the Methods section of the Manuscript.

## Part 1: Disentangling the heterogeneity, calculation of variance explained by each factor in each view
This is done by the 'plotVarianceExplained' function. The resulting plot allows to get an overivew of which factors are active in which view(s). If a factor is active in more than one view, this means that is capturing shared signal (co-variation) between features of the different views.

```{r}
r2 <- plotVarianceExplained(MOFAobject)
```

## Part 2: Characterisation of individual factors
### Inspection of top weighted features in the active views
To get an overview of the weights across all factors in a given view you can use the 'plotWeightsHeatmap' function.
For example, here we plot all weights from all factors in the Mutation data:
```{r}
plotWeightsHeatmap(MOFAobject, "Mutations", factors=1:5, show_colnames=F)
```
Similarly, we can also plot all weights for a single factor using the 'plotWeights' function.
For example, here we plot all weights from Factor 1 in the Mutation data:
```{r}
plotWeights(MOFAobject, view = "Mutations", factor = 1)
```
If you are only interested in looking at the top features you can use the 'plotTopWeights' function.
For example, here we plot the mutations with largest loadings on Factor 1:
```{r}
plotTopWeights(MOFAobject, "Mutations", 1)
```
Features with large weight in a given factor means that they follow the pattern of covariation associated with the factor. For example, if a factor is capturing sex differences, then the largest weights will be associated with features that show sex-specific effects.

From the previous plots, we can clearly see that Factor 1 is associated to IGHV status. Now if we look at the mRNA view, the top features will also be associated to IGHV status:
```{r}
plotTopWeights(MOFAobject, "mRNA", 1)
```

Finally, instead of looking at an the "abstract" weight, it is useful to observe the coordinated heterogeneity of the top features in the original data. This can be done using the 'plotDataHeatmap' function.
```{r}
plotDataHeatmap(MOFAobject, view="mRNA", factor=1, features=20, show_rownames=FALSE)
```

### Feature set enrichment analysis in the active views
Sometimes looking at the loadings of single features can be challenging, and often the combination of signal from functionally related sets of features (i.e. gene ontologies) is required.

Here we implemented a function for feature set enrichment analysis method (FeatureSetEnrichmentAnalysis) derived from the [PCGSE package](https://cran.r-project.org/web/packages/PCGSE/index.html).

The input of this function is a MOFA trained model (MOFAmodel), the factors for which to perform feature set enrichment (a character vector), the feature sets (a binary matrix) and a set of options regarding how the analysis should be performed, see also documentation of 'FeatureSetEnrichmentAnalysis'

We illustrate the use of this function using the [reactome](http://reactome.org) annotations.
```{r}
# Load reactome annotations, binary matrix with feature sets in rows and feautres in columns
data("reactomeGS")

# perfrom enrichment analysis
fsea.out <- FeatureSetEnrichmentAnalysis(MOFAobject, "mRNA", reactomeGS, alpha = 0.01)
```

The next step is to visualise the results of the Gene Set Enrichment Analysis. There are two default plots:
(a) General Overview: Barplot with number of enriched gene sets per view
```{r}
Barplot_FeatureSetEnrichmentAnalysis(fsea.out, alpha=0.01)
```

(b) Factor-specific:
```{r}
interestingFactors <- 4:5
for (factor in interestingFactors) {
  lineplot <- LinePlot_FeatureSetEnrichmentAnalysis(fsea.out, factor, threshold=0.01, max.pathways=10)
  print(lineplot)
}
```


## Ordination of samples by factors to reveal clusters and graadients in the sample space

Samples can be visualized along factors of interest using the 'FactorsScatterPlot' function.
```{r}
plotFactorScatter(MOFAobject, factors = 1:2, color_by = "IGHV", shape_by = "trisomy12")
```

An overview of pair-wise sctterplots for all or a subset of factors is produced by the 'FactorsScatterPairs' function
```{r}
plotFactorScatters(MOFAobject, factors = 1:4, color_by = "IGHV")
```

A single factor can be visualised using the 'FactorBeeswarmPlot' function
```{r}
plotFactorBeeswarm(MOFAobject, factors = 1, color_by = "IGHV")
```

## Customized analysis
For customized exploration of weights and factors, you can directly fetch the variables from the model using 'get' functions: 'getWeights', 'getFactors' and 'getTrainData':
```{r}
MOFAweights <- getWeights(MOFAobject, views="all", factors="all", as.data.frame = T)
head(MOFAweights)
```

```{r}
MOFAfactors <- getFactors(MOFAobject, factors=c(1,2), as.data.frame = F)
head(MOFAfactors)
```

```{r}
MOFAdata <- getTrainData(MOFAobject, as.data.frame = T, views = "Mutations")
head(MOFAdata)
```


# Further functionalities 

## Prediction of views
With the 'predict' function, full views can be predicted based on the MOFA model with all or a subset of factors
```{r}
predictedDrugs <- predict(MOFAobject, view="Drugs", factors="all")
```

## Imputation of missing observations
With the 'imputeMissing' function all missing values are imputed based on the MOFA model. The imputed data is then stored in the ImputedData slot of the MOFAobject and can be accessed via the 'getImputedData' function.
```{r}
MOFAobject <- imputeMissing(MOFAobject)
imputedDrugs <- getImputedData(MOFAobject, view="Drugs")
```

## Clustering of samples based on latent factors
Samples can be clustered according to their values on some or all latent factors using the 'clusterSamples' function. CLusters can for example be visualised using the FactorsScatterPlot function
```{r}
h <- clusterSamples(MOFAobject, k=2, factors=1)
plotFactorScatters(MOFAobject, factors=1:2, color_by=h$cluster)
```


